src/main/java/com/epam/reportportal/spock/AbstractLaunchContext.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2016 EPAM Systems
+ *
+ *
+ * This file is part of EPAM Report Portal.
+ * https://github.com/reportportal/agent-java-spock
+ *
+ * Report Portal is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Report Portal is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Report Portal.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.epam.reportportal.spock;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+
+import javax.annotation.Nullable;
+
+import org.spockframework.runtime.model.FeatureInfo;
+import org.spockframework.runtime.model.IterationInfo;
+import org.spockframework.runtime.model.SpecInfo;
+
+/**
+ * Context which stores and provides the reporting meta data during the test launch
+ *
+ * @author Dzmitry Mikhievich
+ */
+abstract class AbstractLaunchContext {
+
+	private Boolean launchInProgress;
+	private String launchId;
+
+	@Nullable
+	String getLaunchId() {
+		return launchId;
+	}
+
+	void setLaunchId(String launchId) {
+		this.launchId = launchId;
+	}
+
+	/**
+	 * @return true if launch status hadn't been started previously, false
+	 *         otherwise
+	 */
+	synchronized boolean tryStartLaunch() {
```

Comment:
Basically `launchInProgress`  should has 3 states: 

- null -> launch isn't started, 
- true -> launch in progress, 
- false -> launch is finished. 

I've considered `AtomicBoolean` application, but the implementation of the first state looks unsafe: if we'll init this value lazily without any locks on the level  of launch context, potentially it can lead to the race condition. As an option we can use `AtomicReference` to synchronize an access to the  flag...What do you think about it?